### PR TITLE
🐦 Canary non-breaking updates

### LINKED ISSUE #667: Weekly Util is broken
### Description of the problem

Isn't reading historical and isn't saving to database by the looks

### Reproduction of the problem

```bash
...
```

### PR DESCRIPTION
**Pull request by davidzwa**
This PR contains the changes on canary **without awilix**. This is to keep canary somewhat close to dev and make it manageable.

### TL;DR
- Prettier col width 100
- Dependencies installed (check `package.json`)
- Some generics like logger or api exception handler and some files added which are used in the `awilix` branch

### The PR's
1) Weekly Util PR #686 (https://github.com/OctoFarm/OctoFarm/commit/b9072df64089a93e8c1b7336a11a3d7996f336ea)
- Fixed #667 Weekly Utility was not saving previous days values.
2) Squash and fix tests #694
- Cleanup docker files
- Fixes tests and GHA
- Purge 2x legacy folders
- Move and push new non-breaking changes or additions for utils (took it out of dataFunctions folder)
- Printer printerURL prop now required + fix tests
3) Shuffle files, add new unused files #696
- Move some root files to server_src
- Push non-breaking changes in generics (EDIT: just additions, no removal)
4) Add new deps, run prettier on most files #697
- Add websocket dependencies like rxjs, rjxs-websockets, hexnut-client and hexnut-handle (possibly removed later)
- Run prettier with 100 column width (saves loads of lines of code)
5) Remove hexnut-client as we're not going to use it #702 
6) Add global exception handler, fix client some `.json()` dupes and add generics like constants (unused in dev code) #704 
7) Add EventEmitter2 dependency, not used yet. Update with octoprint.js .json() change. Revert v2 view change. #705 
8) Move over local OF client to axios and fix page loading because of unneeded .json() calls. #706 
9) Additional .json() clean up #708 
10) Introduce client side axios #706 #707 
11) Revert Printer model rename without proper rename migration #712 
11) client error handling improvements #711 
12) Slightly reverted some choices in octofarm_client.js #714 
13) Various fixes for canary into dev #728

### CODE REVIEW

**Comment by davidzwa** [comment]
TODO:
.json() dupes with new octofarm_client implementation

File: server_2x/views/layout.ejs [hunk 1]
```diff
@@ -59,12 +59,12 @@
             </li>
             <li class="nav-item">
                 <a title="Davids super power tool" class="nav-link text-light text-center btn-sm d-none"
-                   id="menuListView" href="/mon/printerMap"><i class="fas fa-list"></i> Map</a>
+                   id="menuListView" href="/mon/map"><i class="fas fa-list"></i> Map</a>
```
Comment (NotExpectedYet) [inline server_2x/views/layout.ejs:62]: These need to be added to the changelog please

File: server_2x/views/layout.ejs [hunk 2]
```diff
@@ -59,12 +59,12 @@
...
             </li>
             <li class="nav-item">
                 <a title="Current Operations Monitoring Layout View"
                    class="nav-link text-light text-center btn-sm <%= noPrinters %>" id="menucurrentOperationsView"
-                   href="/mon/currentOp"><i class="fas fa-star"></i>
+                   href="/mon/current-ops"><i class="fas fa-star"></i>
```
Comment (NotExpectedYet) [inline server_2x/views/layout.ejs:67]: This to the changelog as well.

File: server_src/constants/state.constants.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+        date: null
+      },
+      job: {
+        date: null
+      },
+      file: {
+        date: null
+      },
+      status: "danger",
+      date: null
```
Comment (NotExpectedYet) [inline server_src/constants/state.constants.js:94]: I like the idea for this, it seems more of a OctoFarm thing than a printer thing personally.

File: server_src/constants/state.constants.js [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+  }
+  if (state === PSTATE.GlobalAPIKey) {
+    return { name: "danger", hex: "#2e0905", category: CATEGORY["Error!"] };
+  }
+  if (state === PSTATE["Error!"]) {
+    return { name: "danger", hex: "#2e0905", category: CATEGORY["Error!"] };
+  }
+  if (state === PSTATE.Offline) {
+    return { name: "danger", hex: "#2e0905", category: CATEGORY.Offline };
+  }
```
Comment (NotExpectedYet) [inline server_src/constants/state.constants.js:183]: This should be "Disconnected" as before.

File: server_src/constants/state.constants.js [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    return { name: "success", hex: "#00330e", category: CATEGORY.Idle };
+  }
+  if (state === PSTATE["Offline after error"]) {
+    return { name: "danger", hex: "#2e0905", category: CATEGORY["Error!"] };
+  }
+
+  // TODO is this a smart idea? No error?
+  console.warn("PSTATE not recognized:", state);
+  return { name: "warning", hex: "#583c0e", category: CATEGORY.Active };
+};
```
Comment (NotExpectedYet) [inline server_src/constants/state.constants.js:209]: A missed state in my opinion shouldn't stop the printer connection from working to me. I defaulted back to active because it was generally there we're I may have missed a state.

File: server_src/handlers/generic-websocket.adapter.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
+/**
+ * An OctoFarm websocket handler suitable for any websocket application like OctoPrint, Mariner, Klipper, etc.
+ * Copy it to fit your needs, or take a look at 'octoprint-websocket.adapter.js'
+ */
+module.exports = class GenericWebsocketAdapter {
```
Comment (NotExpectedYet) [inline server_src/handlers/generic-websocket.adapter.js:5]: Nice!

File: server_src/migrations/0001_printers_settingsapperance_13072021.migration.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    // Change the schema version if we may
+
+    Printer.updateMany({}, { $rename: { settingsApperance: "settingsAppearance" } });
+    Printer.updateMany({}, { $rename: { apikey: "apiKey" } });
+  }
+
+  down(schemaVersion) {}
+};
+
+module.exports = PrinterSchemaMigrator13072021;
```
Comment (NotExpectedYet) [inline server_src/migrations/0001_printers_settingsapperance_13072021.migration.js:15]: Love this

Comment (NotExpectedYet) [reply server_src/migrations/0001_printers_settingsapperance_13072021.migration.js:15]: Just to check the client knows about the apiKey change?

**Review by NotExpectedYet — CHANGES_REQUESTED** [review]
LGTM just had some questions

**Comment by NotExpectedYet** [comment]
Will test now

File: server_2x/views/layout.ejs (hunk 2 above)
Comment (davidzwa) [reply server_2x/views/layout.ejs:67]: Huh this should not have reached canary. Actually a good catch

File: server_2x/views/layout.ejs (hunk 1 above)
Comment (davidzwa) [reply server_2x/views/layout.ejs:62]: Also this one should not have reached canary yet... stupid me and good catch!

**Comment by NotExpectedYet** [comment]
Some bugs: 

This is when loading list, Panel/Camera. Think you missed a settings.json when converting to the new OF client. 
Uncaught (in promise) TypeError: settings.json is not a function
    checkFilamentManager filamentGrab.js:5

Think this is related to the above. List view doesn't replace the Loading... PanelView/Camera view are fine
image

File: client_src/js/lib/octoprint.js [hunk 8]
```diff
@@ -1,8 +1,18 @@
...
 export default class OctoPrintClient {
+  static validatePrinter(printer) {
+    if (!printer.apikey) {
+      throw new Error("Api key not provided");
+    }
+    if (!printer.printerURL) {
+      throw new Error("Printer URL not provided");
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (davidzwa) [reply client_src/js/lib/octoprint.js:13]: My reasoning is that if the apikey or printerURL are not provided then something bad has happened. My thoughts then are to throw an error to break out of this client's code which should be reported by the user to us. Handling could be done on a global level with an UI alert or an error dialog.
This is on the awilix branch as well and works a treat for me with the new backend exception handler

An anonymous bug reporter tool would be nice to report it to us. What's your thought on such a telemetry solution?

**Comment by davidzwa** [comment]
> Some bugs:
> 
> This is when loading list, Panel/Camera. Think you missed a settings.json when converting to the new OF client.
> Uncaught (in promise) TypeError: settings.json is not a function
> checkFilamentManager filamentGrab.js:5
> 
> Think this is related to the above. List view doesn't replace the Loading... PanelView/Camera view are fine
> image

Lots of duplicate `.json()` calls will be present. Would you be up to fixing that? Its a bit unorthodox, just got my head in the "big one" branch lots of things still to be done there

**Comment by NotExpectedYet** [comment]
I've lost printer names. Still in database. 

image

I don't get a terminal stream 
image

Prints fail to start:
```
Uncaught (in promise) TypeError: checkSettings.json is not a function
    jobAction octoprint.js:246
    applyListeners printerManager.js:1561
    applyListeners printerManager.js:1555
    init printerManager.js:86
    createOrUpdatePrinterTableRow printer-data.js:234
    createOrUpdatePrinterTableRow printer-data.js:231
    createOrUpdatePrinterTableRow printer-data.js:212
    workerEventFunction printer-manager.functions.js:27
    onmessage client-worker.js:13
    webWorkerFunction client-worker.js:11
    createClientSSEWorker client-worker.js:40
    <anonymous> printer-management.runner.js:202
    <anonymous> printer-management.runner.min.js:36018
    <anonymous> printer-management.runner.min.js:36020
```

History doesn't load 
image
```
Uncaught (in promise) TypeError: newHistory.json is not a function
```

Filament Manager doesn't load graphs. 
```
Uncaught (in promise) SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
```

Settings doesn't load graphs
```
Uncaught (in promise) ReferenceError: Client is not defined
    updateStatus settings.js:404
    setInterval handler* settings.js:403
    <anonymous> settings.min.js:5631
    <anonymous> settings.min.js:5633
```

File: server_src/migrations/0001_printers_settingsapperance_13072021.migration.js (hunk 7 above)
Comment (davidzwa) [reply server_src/migrations/0001_printers_settingsapperance_13072021.migration.js:15]: Note: this code isnt run yet. Its something to show "this is how it will be later". I'd be ok with purging it from canary if its better.

In future we will test the migration fully. If the client can be easily updated we should in such cases, if not we should convert it to old props. Got a very neat mapping function in the new code, will show you later. Its basically PrinterClean but then on-the-fly (no calculations, just flattening and mapping of properties to match the client interface)

File: server_src/handlers/generic-websocket.adapter.js (hunk 6 above)
Comment (davidzwa) [reply server_src/handlers/generic-websocket.adapter.js:5]: Still very clunky, just to give an idea. Also the adapter design is currently quite bad, will need to work on that

File: server_src/migrations/0001_printers_settingsapperance_13072021.migration.js (hunk 7 above)
Comment (NotExpectedYet) [reply server_src/migrations/0001_printers_settingsapperance_13072021.migration.js:15]: Nah it's fine to me if it's like that. 

The client thankfully is all in one place where it utilises those keys so it's not a hard update.

File: client_src/js/lib/octoprint.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (NotExpectedYet) [reply client_src/js/lib/octoprint.js:13]: I just didn't want the client locking up but yeah I am thinking of moving over to axois and writing the functions here into a better style so they won't need to be try/caught everywhere and a exception handler can be written into the client. Like what you did with that server side so it's a no-brainer. 

My only issue is where does it report too, Might be me been jaded from clicking that initially in Android and other platforms and never getting a response like but I've always preferred the manual solution. Depends on what specifically and how much is getting reported. Something like an automated/submitted from the app solution to me would fall under paid as we'd need to allocated resources to action it.

**Comment by NotExpectedYet** [comment]
> > Some bugs:
> > This is when loading list, Panel/Camera. Think you missed a settings.json when converting to the new OF client.
> > Uncaught (in promise) TypeError: settings.json is not a function
> > checkFilamentManager filamentGrab.js:5
> > Think this is related to the above. List view doesn't replace the Loading... PanelView/Camera view are fine
> > image
> 
> Lots of duplicate `.json()` calls will be present as `octofarm_client` now does that internally. Would you be up to fixing that? Its a bit unorthodox, just got my head in the "big one" branch lots of things still to be done there

Yeah no trouble.

File: server_src/constants/state.constants.js (hunk 3 above)
Comment (davidzwa) [reply server_src/constants/state.constants.js:94]: That's a good point. Im actually struggling with it in the new architecture. The systemChecks are coupled to each printer on `master`
https://github.com/OctoFarm/OctoFarm/blob/e99cabd3f27f3ab6e21c27f0da5e8e319c4a2751/server_src/lib/dataFunctions/printerClean.js#L596
So in that regard it used to be printer based.... or let me say "device based" if we're abstracting to multiple interfaces

File: server_src/handlers/generic-websocket.adapter.js (hunk 6 above)
Comment (NotExpectedYet) [reply server_src/handlers/generic-websocket.adapter.js:5]: Excuse my ignorance but why bad?

File: server_src/constants/state.constants.js (hunk 3 above)
Comment (NotExpectedYet) [reply server_src/constants/state.constants.js:94]: Yeah sorry this was more of a " should we move this ? " question than a statement. As these are invariably going to to be task based now no? 

The UI could be on System as a little health check and the last run time outputted on the log dump (just as ideas)

File: server_2x/views/layout.ejs (hunk 2 above)
Comment (davidzwa) [reply server_2x/views/layout.ejs:67]: Reverted in #705

File: server_2x/views/layout.ejs (hunk 1 above)
Comment (davidzwa) [reply server_2x/views/layout.ejs:62]: Reverted in #705

File: server_src/constants/state.constants.js (hunk 3 above)
Comment (davidzwa) [reply server_src/constants/state.constants.js:94]: My mind is a bit fried... Ive been struggling with how we process data and my mental solution is to not do it backend but on client side. Im want to make sure that it's not because of my bad design so Ill ponder on it a bit

File: server_src/handlers/generic-websocket.adapter.js (hunk 6 above)
Comment (davidzwa) [reply server_src/handlers/generic-websocket.adapter.js:5]: Bad in the awilix context, so bit hard to explain as I cant put my finger on it. I guess I need the P.O.V. from a potential OF plugin dev to truly understand if this adapter is going to help them.

File: server_src/constants/state.constants.js (hunk 5 above)
Comment (davidzwa) [reply server_src/constants/state.constants.js:209]: I guess we should log this in dev mode only to spot OP API changes we didnt foresee.

Comment (NotExpectedYet) [reply server_src/constants/state.constants.js:209]: Yeah that sounds like a good plan. Seeing the OP file now it shouldn't be too hard to keep up with them anyway. I just think a fall back to offline is a bit confusing to the user. How about we fall back to "unknown" state? Could be ignored then in the rest of the backend / client as nothing is looking for it

File: server_src/constants/state.constants.js (hunk 3 above)
Comment (NotExpectedYet) [reply server_src/constants/state.constants.js:94]: > Answering your question: Im not sure it should be task based as updates are mostly caused by WebSocket messages. I would expect the resulting data from that to be updated by websocket state changes

Job and Printer yeah totally websockets true. 

History and other already in there aha. 

> my mental solution is to not do it backend but on client side. Im want to make sure that it's not because of my bad design so Ill ponder on it a bit.
To me the UI shoud UI for lack of a better description aha. So anything relating to displaying information. The server should be prepping information, ie calculations/settings and what not. There is a rather blatant dis-regard for that at the moment with some html server side, which should go.

File: server_src/constants/state.constants.js (hunk 5 above)
Comment (NotExpectedYet) [reply server_src/constants/state.constants.js:209]: Your right Active is also confusing btw aha

File: server_src/handlers/generic-websocket.adapter.js (hunk 6 above)
Comment (NotExpectedYet) [reply server_src/handlers/generic-websocket.adapter.js:5]: > Bad in the awilix context, so bit hard to explain as I cant put my finger on it. I guess I need the P.O.V. from a potential OF plugin dev to truly understand if this adapter is going to help them.

For me, if it helps, when I'm writing the connectors for the others I want to be able to open up a websocket, pass the listener functions and have the resulting information go through the cleaners. Now an issue I've always thought of is making this information available to the rest of the system, and also the UI changing slightly for each type of web interface. 

I'd gather we'd have a preset set of variables anyone using the adapters must stick to to be in keeping with the current backend. New variables could be added / removed based on the type of printer.  

Resin printers won't need temp, and half the control options for example. 
Klipper the file stuff would be slightly different (I think)

File: server_src/constants/state.constants.js (hunk 5 above)
Comment (davidzwa) [reply server_src/constants/state.constants.js:209]: Nice I will make sure to pick this up in le big one

File: client_src/js/lib/octoprint.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (davidzwa) [reply client_src/js/lib/octoprint.js:13]: My experience with anonymous bug tracking is that its: optional and unpersonal. I envisioned it as ticket-less telemetry to only help us.
The resource question is then easier to answer... if the bugtracking endpoint is overwhelmed well we got work to then xD

**Comment by NotExpectedYet** [comment]
@davidzwa  

I've migrated to axios and fixed the calls/cleaned up where necessary. (think i got it all) 

I'm currently happy for this to enter development now.

File: client_src/js/lib/octoprint.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (NotExpectedYet) [reply client_src/js/lib/octoprint.js:13]: > My experience with anonymous bug tracking is that its: optional and unpersonal. I envisioned it as ticket-less telemetry to only help us better, which in effect helps users - just like we do now.
> The resource question is then easier to answer... if the bugtracking endpoint is overwhelmed well we got work to then xD

If it's not user controlled and only fired on an actual error somewhere we can track/trace personally then I'm game for that totally. I was stuck on you wanted a user to press submit xD

**Review by NotExpectedYet — COMMENTED** [review]
Infact I'm going to have a double look through just to be sure.

**Review by davidzwa — COMMENTED** [review]
OF Client with axios could do with more specific error handling. This is aiming at the fact that we'll be handling exceptions more predictable on backend soon with "Le Big One"

File: CHANGELOG.md [hunk 9]
```diff
@@ -19,6 +20,11 @@ All notable changes to this project will be documented in this file.
     - Refactor monitoring pages (panel, list, camera) with latest SSE client reusability.
     - Refine the printer map layout with borders and printer quick actions (page not public yet).
     - Reduced logging (we will bring it back later in different shape with an Exception handler)
+    - Replaced fetch with axios in the local OctoFarm client calls.  
+    - Moved all error handling for client connections into axios calls
+    - Refactored System page into seperate manageable files ejs/js, cleaned up a lot of code. 
+    - Updated the system page layout. 
+    - Moved filament manager plugin actions to serperate service file.
```
Comment (davidzwa) [inline CHANGELOG.md:27]: Serparate typo

File: CHANGELOG.md [hunk 10]
```diff
@@ -19,6 +20,11 @@ All notable changes to this project will be documented in this file.
     - Refactor monitoring pages (panel, list, camera) with latest SSE client reusability.
     - Refine the printer map layout with borders and printer quick actions (page not public yet).
     - Reduced logging (we will bring it back later in different shape with an Exception handler)
+    - Replaced fetch with axios in the local OctoFarm client calls.  
+    - Moved all error handling for client connections into axios calls
+    - Refactored System page into seperate manageable files ejs/js, cleaned up a lot of code.
```
Comment (davidzwa) [inline CHANGELOG.md:25]: Separate typo

Comment (NotExpectedYet) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md (hunk 9 above)
Comment (NotExpectedYet) [reply CHANGELOG.md]: Fixed

**Review by NotExpectedYet — APPROVED** [review]
LGTM

**Comment by davidzwa** [comment]
We gotta fix the SSE goof

**Comment by davidzwa** [comment]
@NotExpectedYet my printer Control terminal is empty. Looking back you had this as well. Any clue how to fix it?

**Comment by NotExpectedYet** [comment]
@davidzwa yeah I did, didn't check that. It comes down the websocket from printerClean but I'll do an inspect.

**Comment by NotExpectedYet** [comment]
I am finishing early today so I'll spin it up and have a test

**Comment by NotExpectedYet** [comment]
@davidzwa Will have a look at this now

**Comment by davidzwa** [comment]
> Tested each section seems good to go now.

Nice, feel free to merge
